web/packages/core/src/ruffle-player.ts

```diff
@@ -149,6 +152,8 @@ export class RufflePlayer extends HTMLElement {
     private _cachedDebugInfo: string | null = null;
 
     private isExtension = false;
+    private longPressTimer: ReturnType<typeof setTimeout> | null = null;
```

Comment:
Actually, now it seems like `setTimeout` is not need at all, since we only actively check whether the given timeout has elapsed or not, rather than passively waiting for a callback to be triggered.
So, instead we can simply store the last timestamp a `pointdown` occurred (via `Date.now()`), then in `checkLongPress()` compare it against the current timestamp to tell whether enough time has elapsed.

Reply:
On a second thought maybe showing the context menu on `pointerup` can be inconvenient. It makes sense with a right-click, but with touch the user might not know how long they need to keep touching. Also context menu on Android is shown when still touching (and I believe on iOS as well).
If so, maybe we should return back to `pointerdown` and which case you can scratch my previous comment.

Reply:
> So, instead we can simply store the last timestamp a `pointdown` occurred (via `Date.now()`), then in `checkLongPress()` compare it against the current timestamp to tell whether enough time has elapsed.

I'll be honest, I prefer setTimeout, even without a callback. Time comparisons in JavaScript are finicky, while `setTimeout` is a well-documented standard function of the language.

Also, the contextmenu behavior is the exact behavior that every context menu follows on Android, from popping up after a little under a second to appearing before releasing touch, everything is exactly as it is with every other context menu functionality, and therefore exactly matches what an Android user expects. The idea here is just to match that behavior with the behavior on an iPhone. Popping up the context menu on any touch (pointerdown) without waiting severely messes with the functionality of touch-based games. Needing to hide the context menu to do anything on touch-based games on mobile severely detracts from the usefulness of the options within it.

Reply:
Also, I think clearTimeout is needed so it can be cancelled when touch is released before the elapsed time has passed.

Reply:
@relrelb Here is my attempt to make the iOS context menu work like it does on Android (on pointerdown instead of pointerup): https://github.com/n0samu/ruffle/commit/4a655bf1be9251c6684597cb2a10120c3c0c4e39
With the context menu appearing on the pointerdown event, now there is no way to stop the pointerup event from propagating and immediately closing the context menu. The only solution I found was to add an event listener, which itself adds another event listener that finally closes the context menu. That way the context menu closes after the second pointerup event that is received, not the first (which fires immediately after the user lifts their finger).

However, I think this is a considerably worse user experience and would take far too much effort to improve. The context menu appears where the user's touch began, since that is the location of the pointerdown event. If we wanted the context menu to appear in the right place, we would need to store the user's current finger position and update it on every pointermove event. And then we have the problem of detecting "long presses": how do we detect that the user's finger moved "too far" for this to be a long press? It's pretty annoying for the context menu to appear while you're sliding your finger. It's comparatively less annoying for it to appear after you lift your finger (though it's still not ideal).

So can we please leave things as-is and proceed with merging this PR?

Reply:
@n0samu Good arguments - I wasn't aware of them and they do make sense.
Indeed simulating UI logic can be an headache, but IIUC even under the current approach, if the user drags their finger "too far" for at least 800ms, then a context menu will be shown as well. The only difference is that this will happen when the finger is lifted, am I right?
So considering this, it seems like the "too far" case should be detected either way. I think this can implemented relatively easily - In `startLongPressTimer()` store the starting coordinates, then in `checkLongPress()` compare the distance between the starting and final coordinates against some threshold. If it's too far - don't show a context menu.
However this is only a partial implementation, since a context menu would still be shown when the finger is dragged a long distance but lifed just where it started... To handle this case the `pointermove` event should be used like you mentioned.

Which approach should be taken? I'm not sure. We just need to decide which is "good enough".

Reply:
> The only difference is that this will happen when the finger is lifted, am I right?

Yes that's right.

As for the "too far" behavior, I'm not familiar enough with Android to say how it works or if it even works that way at all. And tuning this behavior would require some experimentation. I'm not really up for all that.

Reply:
I'll try to summarize the conditions for a context menu on Android:
1. It is shown on `pointerdown` (doesn't wait for the finger to be lifted).
2. The finger must be kept "in place" for about 0.5-1 seconds.
3. If the finger is moved sooner, then the underlying text is selected.

I don't have an available iOS device, but I believe it behaves the same.
Anyway, I guess the current logic is fine enough, given that changing it further is going to be complex. Worst-case we'll get feedback from users and tune it accordingly.

So besides my 2 comments below, I see this PR ready to be merged.

Reply:
Yeah, as nosamu mentioned switching from pointerup to pointerdown is complex. I also noticed the timing being somewhere between 0.5 and 1 seconds, which is why I guesstimated 0.8 seconds. And pointermove appears to fire at the slightest amount of movement, even the natural resting movement of your finger, so that's what makes the too far definition complex. I'll address the new comments in the next few hours.